tracing/traced_file.cc

```diff
@@ -0,0 +1,185 @@
+    trace("call write_dma(pos = {}, std::vector<iovec>, const io_priority_class&)", pos);
+    return _f_impl.write_dma(pos, std::move(iov), pc).then([this, pos] (size_t ret) {
+        trace("ret from write_dma(pos = {}, std::vector<iovec>, const io_priority_class&) = {}", pos, ret);
+        return ret;
+    });
+}
+
+future<size_t>
+traced_file_impl::read_dma(uint64_t pos, void* buf, size_t len, const io_priority_class& pc) {
+    trace("call read_dma(pos = {}, len = {}, const io_priority_class&)", pos, len);
+    return _f_impl.read_dma(pos, buf, len, pc).then([this, pos, len] (size_t ret) {
+        trace("ret from read_dma(pos = {}, len = {}, const io_priority_class&) = {}", pos, len, ret);
+        return ret;
+    });
+}
+
+future<size_t>
+traced_file_impl::read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) {
+    trace("call read_dma(uint64_t pos = {}, std::vector<iovec> iov, const io_priority_class&)", pos);
+    return _f_impl.read_dma(pos, std::move(iov), pc).then([this, pos] (size_t ret) {
+        trace("ret from read_dma(uint64_t pos = {}, std::vector<iovec> iov, const io_priority_class&) = {}", pos, ret);
+        return ret;
+    });
+}
+
+future<>
+traced_file_impl::flush(void) {
+    trace("call flush()");
+    return _f_impl.flush().then([this] {
+        trace("ret from flush()");
+    });
+}
+
+future<struct stat>
+traced_file_impl::stat(void) {
+    trace("call stat()");
+    return _f_impl.stat().then([this] (struct stat s) {
+        trace("ret from stat()");
+        return s;
+    });
+}
+
+future<>
+traced_file_impl::truncate(uint64_t length) {
+    trace("call truncate(length = {})", length);
+    return _f_impl.truncate(length).then([this, length] {
+        trace("ret from truncate(length = {})", length);
+    });
+}
+
+future<>
+traced_file_impl::discard(uint64_t offset, uint64_t length) {
+    trace("call discard(offset = {}, length = {})", offset, length);
+    return _f_impl.discard(offset, length).then([this, offset, length] {
+        trace("ret from discard(offset = {}, length = {})", offset, length);
+    });
+}
+
+future<>
+traced_file_impl::allocate(uint64_t position, uint64_t length) {
+    trace("call allocate(position = {}, length = {})", position, length);
+    return _f_impl.allocate(position, length).then([this, position, length] {
+        trace("ret from allocate(position = {}, length = {})", position, length);
+    });
+}
+
+future<uint64_t>
+traced_file_impl::size(void) {
+    trace("call size()");
+    return _f_impl.size().then([this] (uint64_t ret) {
+        trace("ret from size() = {]", ret);
+        return ret;
+    });
+}
+
+future<>
+traced_file_impl::close() {
+    trace("call close()");
+    return _f_impl.close().then([this] {
+        trace("ret from close()");
+    });
+}
+
+std::unique_ptr<seastar::file_handle_impl>
+traced_file_impl::dup() {
+    return _f_impl.dup();
```

Comment:
Is it fine to unwrap file here? One thing that I'm afraid of is that we may accidentally stop tracing if the code path goes through dup().

Reply:
It's documented:
```
// Note: calling dup() on the wrapper returns a handle to the underlying file.
seastar::file make_traced_file(seastar::file f, trace_state_ptr trace_state, seastar::sstring trace_prefix);
```

implementing `dup` so that it returns a wrapped handle would require writing another `file_handle_impl` subclass, which is a lot more work with little gain.
The caller can instead wrap the new file using `make_traced_file` again.

The same approach is taken in `tracking_file_impl` and `checked_file_impl`.